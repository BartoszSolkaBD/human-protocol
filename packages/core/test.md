1. Can we use viaIR with foundry?

Contracts are compuling with viaIR using Foundry. 
You specify via-ir=true in Foundry.toml

2. If we start using viaIR, can we still upgrade deployed proxies?

Contracts were upgraded using viaIR in fact using Foundry. 


3. If we use viaIR, how will it affect to the gas cost?

Using viaIR, gas cost decreased slightly. 
Use : forge snapshot --diff with and without via-ir=true in Foundry.toml


4. If we move to Foundry, can we still upgrade deployed proxies?

Yes, deployed proxies were upgraded using Foundry. 