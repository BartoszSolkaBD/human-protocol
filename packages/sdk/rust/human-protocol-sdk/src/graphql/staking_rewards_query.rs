use serde::Serialize;


mod schema {
  cynic::use_schema!("schemas/reward.graphql");
}


#[derive(Serialize, Debug, Clone, cynic::QueryFragment)]
#[cynic(schema_path = "schemas/reward.graphql")]
pub struct Reward {
  escrow_address: String,
  staker: String,
  slasher: Option<String>,
  amount: cynic::Id
}

#[derive(cynic::QueryVariables, Clone)]
pub struct RewardFilter {
  pub slasher: String
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Root", variables = "RewardFilter")]
#[cynic(schema_path = "schemas/reward.graphql")]
pub struct GetRewardsQuery {
  #[arguments(
    where: { 
      slasher_contains: $slasher
    }
  )]
  pub reward_added_events: Option<Vec<Reward>>,
}

pub async fn run_query(url: &str, filter: &RewardFilter) -> cynic::GraphQlResponse<GetRewardsQuery> {
  use cynic::http::ReqwestExt;

  let query = build_query(filter);
  
  reqwest::Client::new()
    .post(url)
    .run_graphql(query)
    .await
    .unwrap()
}

fn build_query(filter: &RewardFilter) -> cynic::Operation<GetRewardsQuery, RewardFilter> {
  use cynic::QueryBuilder;

  GetRewardsQuery::build(filter.clone())
}

#[cfg(test)]
mod test {
    use crate::{enums::ChainId, constants::NETWORKS};

    use super::*;
    use tokio::test;

    #[test]
    async fn snapshot_test_query() {
      let filter = RewardFilter {
        slasher: "some_address".to_string()
      };
      let query = build_query(&filter);

      insta::assert_snapshot!(query.query);
    }

    #[tokio::test]
    async fn test_running_query() {
      let network = NETWORKS.get(&ChainId::PolygonMumbai).unwrap();
      let filter = RewardFilter {
        slasher: "0x1234567890123456789012345678901234567890".to_string()
      };
      let result = run_query(network.subgraph_url, &filter).await;
      if let Some(errors) = &result.errors {
          assert_eq!(errors.len(), 0);
      }
      insta::assert_debug_snapshot!(result.data);
    }

    #[tokio::test]
    async fn test_filter_builder() {
        let filter = RewardFilter {
          slasher: "some_address".to_string()
        };

        assert_eq!(filter.slasher, "some_address".to_string());
    }
}