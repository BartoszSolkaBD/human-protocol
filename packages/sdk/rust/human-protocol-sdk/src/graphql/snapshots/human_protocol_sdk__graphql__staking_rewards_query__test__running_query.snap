---
source: src/graphql/staking_rewards_query.rs
expression: result.data
---
Some(
    GetRewardsQuery {
        reward_added_events: Some(
            [],
        ),
    },
)
